File: api/src/main/java/com/bindglam/neko/api/pack/PackZipper.java
package com.bindglam.neko.api.pack;

import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.file.Files;

public interface PackZipper extends AutoCloseable {
    void addFile(String path, PackFile file);

    default void addFile(File file) {
        try {
            addFile(file.getName(), new PackFile(() -> {
                try(FileInputStream stream = new FileInputStream(file)) {
                    return stream.readAllBytes();
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            }, Files.size(file.toPath())));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    void addDirectory(File directory);

    @Nullable PackFile file(String path);

    void build();
}

File: core/src/main/kotlin/com/bindglam/neko/pack/PackZipper.kt
package com.bindglam.neko.pack

import com.bindglam.neko.manager.PackManagerImpl
import com.bindglam.neko.utils.createIfNotExists
import com.bindglam.neko.utils.getRelativePath
import com.bindglam.neko.utils.listFilesRecursively
import com.bindglam.neko.utils.parallelIOThreadPool
import com.bindglam.neko.utils.toPackFile
import java.io.File
import java.io.FileOutputStream
import java.nio.file.Files
import java.util.concurrent.atomic.AtomicInteger
import java.util.zip.ZipEntry
import java.util.zip.ZipOutputStream

class PackZipper(private val buildFile: File) {
    private val parallelThreadPool = parallelIOThreadPool()

    private val folders = hashSetOf<String>()
    private val entries = hashMapOf<String, PackFile>()

    fun addFile(name: String, file: PackFile) {
        entries[name] = file
    }

    fun addFile(file: File) {
        addFile(file.name, file.toPackFile())
    }

    fun addDirectory(folder: File) {
        folder.listFilesRecursively().forEach {
            if(it.parentFile.path != folder.path)
                folders.add(it.parentFile.getRelativePath(folder.path, "/"))

            entries[it.getRelativePath(folder.path, "/")] = it.toPackFile()
        }
    }

    fun build(onDone: Runnable) {
        buildFile.createIfNotExists()

        ZipOutputStream(FileOutputStream(buildFile)).apply {
        }.use { zipStream ->
            folders.forEach { path ->
                zipStream.putNextEntry(ZipEntry("${path}/"))
                zipStream.closeEntry()
            }

            var cnt = AtomicInteger()

            parallelThreadPool.forEachParallel(entries.entries.toList(), { it.value.size }) { entry ->
                val bytes = entry.value.bytes()

                synchronized(zipStream) {
                    zipStream.putNextEntry(ZipEntry(entry.key))
                    zipStream.write(bytes)
                    zipStream.closeEntry()

                    if(cnt.incrementAndGet() >= entries.size) {
                        onDone.run()
                    }
                }
            }
        }
    }
}